pkg/controller/util_test.go

```diff
@@ -0,0 +1,175 @@
+package controller
+
+import (
+	"context"
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	gatewayv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+	"sigs.k8s.io/sig-storage-lib-external-provisioner/v8/controller"
+)
+
+// newRefGrantList returns a new ReferenceGrant with given attributes
+func newRefGrantList(name, fromNamespace, fromGrop, fromKind, toNamespace, toGroup, toKind, toName string, settoName bool) []*gatewayv1beta1.ReferenceGrant {
+	ReferenceGrantList := []*gatewayv1beta1.ReferenceGrant{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: fromNamespace,
+			},
+			Spec: gatewayv1beta1.ReferenceGrantSpec{
+				From: []gatewayv1beta1.ReferenceGrantFrom{
+					{
+						Group:     gatewayv1beta1.Group(fromGrop),
+						Kind:      gatewayv1beta1.Kind(fromKind),
+						Namespace: gatewayv1beta1.Namespace(toNamespace),
+					},
+				},
+				To: []gatewayv1beta1.ReferenceGrantTo{
+					{
+						Group: gatewayv1beta1.Group(toGroup),
+						Kind:  gatewayv1beta1.Kind(toKind),
+					},
+				},
+			},
+		},
+	}
+
+	if settoName {
+		objectName := gatewayv1beta1.ObjectName(toName)
+		ReferenceGrantList[0].Spec.To[0].Name = &objectName
+	}
+
+	return ReferenceGrantList
+}
+
+func TestIsGranted(t *testing.T) {
+	ctx := context.Background()
+	var requestedBytes int64 = 1000
+	snapapiGrp := "snapshot.storage.k8s.io"
+	snapapiKind := "VolumeSnapshot"
+	coreapiGrp := ""
+	pvcapiKind := "PersistentVolumeClaim"
+	anyvolumedatasourceapiGrp := "hello.k8s.io/v1alpha1"
+	anyvolumedatasourceapiKind := "Hello"
+	fromNamespace := "ns1"
+	fromsrcName := "test-dataSource"
+	toNamespace := "ns2"
+
+	type testcase struct {
+		expectErr    bool
+		volOpts      controller.ProvisionOptions
+		refGrantList []*gatewayv1beta1.ReferenceGrant
+	}
+	testcases := map[string]testcase{
+		"Allowed to access dataSource for xns PVC with refGrant": {
+			volOpts:      generatePVCForProvisionFromXnsdataSource(toNamespace, fromsrcName, "", pvcapiKind, &fromNamespace, &coreapiGrp, requestedBytes, "", true),
+			refGrantList: newRefGrantList("refGrant1", fromNamespace, coreapiGrp, pvcapiKind, toNamespace, coreapiGrp, pvcapiKind, "", false),
```

Comment:
I know we do this in our other test files, but I think it's a little difficult to read the test case with this style of constructor because I need to remember the ordering of the arguments and which field they correspond to. It's perhaps more lines, but I think it's more readable to directly reference the object in the test case definition.

If you want to reduce the boilerplate, maybe you can still have a constructor, but have it take a `ReferenceGrantFrom` and a `ReferenceGrantTo` as arguments?

Similar comment for `generatePVC...` Keep the constructor to reduce on the common boilerplate, but have it take `DataSourceRef` type as an input, which is what changes across each test case.

Reply:
util_test.go created a base test case. And I modified the base test case and modified it to add the test case.
Along with the above modifications, TestProvisionFromSnapshot and TestProvisionFromPVC have also been modified.

In util.go I added the following just in case:
https://github.com/kubernetes-csi/external-provisioner/pull/805/files#diff-ded124c02add3467ce103641d0cd28436ea69cf37c5906470b3d519009d950b7R16-R20

As it turned out to be necessary in the test case below.
https://github.com/kubernetes-csi/external-provisioner/pull/805/files#diff-640957054096359bf678ef9f2789673a60814fba8286165ad298f5eba881536aR166-R170

This condition is unnecessary in the controller because it is acquired by specifying the namespace, but I added it just in case.
https://github.com/kubernetes-csi/external-provisioner/pull/805/files#diff-243ebed2765f75e6a54f57167212fefb08c3b2a85967ad2acbc0eb78919019c1R1936-R1937